app/components/emoji_keyboard.js

```diff
@@ -0,0 +1,203 @@
+    }
+
+    static emptyLayout = () => ({layout: {}})
+    static emojiRange = [...(new Array(emojiCategories.length)).keys()]
+
+    static getCategoryNameByIndex = (index) => {
+        const nonNegativeIndex = Math.max(0, index);
+        const bounded = Math.min(emojiCategories.length - 1, nonNegativeIndex);
+        return emojiCategories[bounded].key;
+    }
+
+    state = {
+        layout: EmojiKeyboard.emojiRange.map(EmojiKeyboard.emptyLayout),
+        currentCategoryName: EmojiKeyboard.getCategoryNameByIndex(0)
+    }
+
+    onEmojiSelected = (emojiName) => {
+        KeyboardRegistry.onItemSelected('EmojiKeyboard', {emojiName});
+    }
+
+    scrollToCategoryByIndex = (index) => {
+        if (index === emojiCategories.length - 1) {
+            this.categoryListRef.scrollToEnd();
+        } else {
+            this.categoryListRef.scrollToIndex({index});
+        }
+    }
+
+    handleCategoryListRef = (ref) => {
+        this.categoryListRef = ref;
+    }
+
+    renderEmoji = (item, index) => (
+        <TouchableOpacity
+            key={index}
+            style={{padding: EMOJI_PADDING}}
+            onPress={() => this.onEmojiSelected(item)}
+        >
+            <Emoji
+                size={EMOJI_SIZE}
+                store={store}
+                key={item}
+                emojiName={item}
+            />
+        </TouchableOpacity>
+    )
+
+    handleCategoryLayout = (index, layout) => {
+        this.setState(
+            {
+                layout: [
+                    ...this.state.layout.slice(0, index),
+                    {layout},
+                    ...this.state.layout.slice(index + 1, this.state.layout.length)
+                ]
+            }
+        );
+    }
+
+    getCategoryLayout = (data, index) => {
+        const categoryData = this.state.layout[index];
+        const {layout} = categoryData;
+        const {width = 0} = layout;
+        return {
+            length: width,
+            offset: this.calculateCategoryOffset(index),
+            index
+        };
+    }
+
+    calculateCategoryOffset = (index) => {
+        return this.state.layout.slice(0, index).reduce(
+            (offset, {layout}) => offset + layout.width
+            , 0
+        );
+    }
+
+    getCategoryNameByIndex = (index) => {
+        return EmojiKeyboard.getCategoryNameByIndex(index);
+    }
+
+    getCategoryNameByOffset = (offset, index = 0) => {
+        if (offset < 0) {
+            return this.getCategoryNameByIndex(index - 1);
+        }
+        const categoryData = this.state.layout[index];
+        return this.getCategoryNameByOffset(
+            offset - categoryData.layout.width,
+            index + 1);
+    }
+
+    handleScroll = ({nativeEvent}) => {
+        const currentScrollOffset = nativeEvent.contentOffset.x;
+        const currentCategoryName = this.getCategoryNameByOffset(currentScrollOffset);
+        this.setState({currentCategoryName});
```

Comment:
:+1 There are two ways: Avoid unwanted calls to `setState` or Implement `shouldComponentUpdate `. I'll take the first approach.